### PR TITLE
[WIP] MV query optimization

### PR DESCRIPTION
**Pull request by jainxrohit**

### CODE REVIEW

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java [hunk 1]
```diff
@@ -1031,6 +1035,58 @@ public void testVirtualBucketing()
...
+                    "GROUP BY ds");
+
+            assertTrue(getQueryRunner().tableExists(getSession(), "test_lineitem_mv"));
+            String explainPlan = getExplainPlan("SELECT _discount_multi_extendedprice_ from test_lineitem_mv", LOGICAL);
+            //TODO: Where is mv gone from the logical plan?
+            System.out.println(explainPlan);
+
+            //TODO: Fails as the scope has a field with empty name
+            explainPlan = getExplainPlan("SELECT SUM(_discount_multi_extendedprice_) from test_lineitem_mv", LOGICAL);
+            System.out.println(explainPlan);
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java:1081]: The failure is expected. I don't think there is anything wrong for this approach. Note that query `SELECT SUM(_discount_multi_extendedprice_) from test_lineitem_mv` doesn't have an output name. Check my comment below.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 2]
```diff
@@ -1143,6 +1120,88 @@ protected Scope visitTable(Table table, Optional<Scope> scope)
...
+            analysis.registerNamedQuery(table, query);
+
+            analysis.registerTableForMV(table);
+            RelationType descriptor = analyzeView(query, name, view.getCatalog(), Optional.of(view.getSchema()), view.getOwner(), table);
+            analysis.unregisterTableForMV(table);
+
+            RelationType outputDescriptor = analysis.getOutputDescriptor(query);
+            Collection<Field> visibleFields = outputDescriptor.getVisibleFields();
+
+            List<ViewDefinition.ViewColumn> columns = visibleFields.stream()
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1187]: Given query `SELECT SUM(_discount_multi_extendedprice_) from test_lineitem_mv` doesn't have an output name, the output visible fields won't provide the name thus will fail.

**Review by highker — COMMENTED** [review]
Overall it's on the right direction. Just need more cleanup on the query construction.

**Comment by jainxrohit** [comment]
> Overall it's on the right direction. Just need more cleanup on the query construction.

Yeah it super messy at the moment, I was planning to make it work first and then cleanup.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 2 above)
Comment (jainxrohit) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1187]: > Given query `SELECT SUM(_discount_multi_extendedprice_) from test_lineitem_mv` doesn't have an output name, the output visible fields won't provide the name thus will fail.

Interesting, are you saying users would need to have it as 
SELECT SUM(_discount_multi_extendedprice_) as _discount_multi_extendedprice_ from test_lineitem_mv to make it work? 
My setup is broken so can not try it at the moment.

Comment (highker) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: not necessary; it's just we don't have an output field name; but the field is still there.

**Pull request closed without merging**